### PR TITLE
[MM-16389] Add emoji icon for profile pictures from webhooks

### PR DESCRIPTION
**Pull request by krjn**
An icon_emoji parameter was added to incoming webhook requests, that overrides the profile icon and the icon_url parameter.

Fixes https://github.com/mattermost/mattermost-server/issues/11194

https://mattermost.atlassian.net/browse/MM-16389

Related PRs:
https://github.com/mattermost/mattermost-server/pull/11586
https://github.com/mattermost/mattermost-mobile/pull/2961
https://github.com/mattermost/mattermost-redux/pull/876
https://github.com/mattermost/docs/pull/2832
https://github.com/mattermost/mattermost-developer-documentation/pull/321

### CODE REVIEW

File: components/post_profile_picture/post_profile_picture.jsx [hunk 1]
```diff
@@ -70,16 +79,21 @@ export default class PostProfilePicture extends React.PureComponent {
...
+
         if (isSystemMessage && !compactDisplay && !fromWebhook && !isBot) {
             return <MattermostLogo className='icon'/>;
         }
-
         const fromAutoResponder = PostUtils.fromAutoResponder(post);
 
         const hasMention = !fromAutoResponder && !fromWebhook;
-        const src = this.getProfilePicSrcForPost(fromAutoResponder, fromWebhook);
+        const profileSrc = this.getProfilePicSrc(fromAutoResponder, fromWebhook);
```
Comment (lieut-data) [inline components/post_profile_picture/post_profile_picture.jsx]: Looks like we don't accept `fromAutoResponder` nor `fromWebhook` in `getProfilePicSrc` -- can we remove?

File: components/post_profile_picture/post_profile_picture.jsx [hunk 2]
```diff
@@ -61,6 +54,22 @@ export default class PostProfilePicture extends React.PureComponent {
         return this.props.status;
     };
 
+    getPostIconURL = (defaultURL, fromAutoResponder, fromWebhook) => {
+        const {enablePostIconOverride, hasImageProxy, post} = this.props;
+        const postIconOverrideURL = post.props.override_icon_url;
+        const useUserIcon = post.props.use_user_icon;
+
+        if (enablePostIconOverride) {
+            if (postIconOverrideURL && postIconOverrideURL !== '') {
```
Comment (lieut-data) [inline components/post_profile_picture/post_profile_picture.jsx]: This logic appears to be subtly different, in that we now allow the override without checking `!fromAutoResponder && fromWebhook`. Can you elaborate on this requirement? I'm guessing it arises from the need to support the emoji, but at the same time allows a case for non-emoji that was previously excluded.

@hmhealey, do you happen to know why these constraints were previously imposed?

File: components/profile_picture.jsx [hunk 3]
```diff
@@ -34,6 +37,12 @@ export default class ProfilePicture extends React.PureComponent {
     }
 
     render() {
+        const profileSrc = (typeof this.props.profileSrc === 'string' && this.props.profileSrc !== '') ?
```
Comment (lieut-data) [inline components/profile_picture.jsx]: Can you elaborate on the need to distinguish `profileSrc` and `src`? I see they are both used below, but I can't quite figure out why we need two.

Comment (krjn) [reply components/profile_picture.jsx]: `profileSrc` will be the original profile picture, and if not overriden, will equal to `src`. The need for this is for the popup of the profile to show the actual profile picture regardless of the post having an overriden icon. This way the 'identity' shown in the popup is of the actual user.

File: components/post_profile_picture/post_profile_picture.jsx (hunk 2 above)
Comment (krjn) [reply components/post_profile_picture/post_profile_picture.jsx]: Nice catch. The behaviour change was not intentional. I reverted and refactored, and will do some more testing.

File: components/profile_picture.jsx (hunk 3 above)
Comment (lieut-data) [reply components/profile_picture.jsx]: Ah, subtle -- thank you! Perhaps if we used terms like `originalProfileSrc` and `profileSrc`? 0/5 on the right terms.

Comment (krjn) [reply components/profile_picture.jsx]: I feel this to be a case where a comment in the code is deserved, so I added one. I think a variable name change would not be enough.

File: components/post_profile_picture/post_profile_picture.jsx [hunk 4]
```diff
@@ -70,16 +78,21 @@ export default class PostProfilePicture extends React.PureComponent {
...
-
         const fromAutoResponder = PostUtils.fromAutoResponder(post);
 
         const hasMention = !fromAutoResponder && !fromWebhook;
-        const src = this.getProfilePicSrcForPost(fromAutoResponder, fromWebhook);
+        const profileSrc = this.getProfilePictureURL();
+        const src = this.getPostIconURL(profileSrc, fromAutoResponder, fromWebhook);
+
+        // if it's relative we assume it's a link to an emoji
+        const isEmoji = src.startsWith('/') && src !== Constants.DEFAULT_WEBHOOK_LOGO;
```
Comment (hmhealey) [inline components/post_profile_picture/post_profile_picture.jsx]: Same comment as on mobile. I don't think we can safely make this assumption with plugins being able to host their own profile pictures

File: components/post_profile_picture/post_profile_picture.jsx [hunk 5]
```diff
@@ -70,16 +79,20 @@ export default class PostProfilePicture extends React.PureComponent {
...
         }
-
         const fromAutoResponder = PostUtils.fromAutoResponder(post);
 
         const hasMention = !fromAutoResponder && !fromWebhook;
-        const src = this.getProfilePicSrcForPost(fromAutoResponder, fromWebhook);
+        const profileSrc = this.getProfilePictureURL();
+        const src = this.getPostIconURL(profileSrc, fromAutoResponder, fromWebhook);
+
+        const isEmoji = post.props && !isEmpty(post.props.override_icon_emoji);
```
Comment (hmhealey) [inline components/post_profile_picture/post_profile_picture.jsx]: Can you just check if `post.props.override_icon_emoji !== ''` here? We don't really need to start using lodash just for this

**Comment by hmhealey** [comment]
@amyblais FYI, the rest of this feature has been merged for 5.14, so we might want to merge this for 5.14 as well. I'm only 0/5 on that though since the main reason for this PR is to make it so that emojis appear a bit nicer when used to override a profile picture. The rest of the feature still functions fine without this in the release as well.

**Comment by amyblais** [comment]
@hmhealey I guess it's fine to merge this for 5.14 as an exception.

**Comment by hanzei** [comment]
@hmhealey Can you please help with the cherry pick?

**Pull request merged**